### PR TITLE
Adding support for PullBaseImageStep to search a manifest list using a user specified os/arch pair

### PR DESCRIPTION
**Pull request by louismurerwa**
This PR adds support for a user to specify the platform he/she wants to use when building an image

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 1]
```diff
@@ -260,14 +265,22 @@ private Image pullBaseImage(
...
    */
-  private ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
+  @VisibleForTesting
+  ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
       RegistryClient registryClient, V22ManifestListTemplate manifestListTemplate)
       throws RegistryException, IOException {
 
-    List<String> digests = manifestListTemplate.getDigestsForPlatform("amd64", "linux");
+    if (buildContext.getContainerConfiguration() != null) {
+      Set<Platform> platforms = buildContext.getContainerConfiguration().getPlatforms();
```
Comment (louismurerwa) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:277]: I was thinking of checking if the platforms set is empty here but i realized that our front end `ContainerBuildPlan ` does the job for us.Should I add a check or should I trust our api?

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:277]: Good question. Because `ContainerBuildPlan` never returns an empty list and we have added a test to ensure it's non-empty, in this case, I think I'm fine with assuming it's non-empty _given that `getContainerConfiguration()` is not null_.

We can certainly add more code to guard against an empty set and raise an error which may also be a good practice in general (depends on situations). But the integration between `ContainerBuildPlan` and the jib-core is pretty tight and and internal, I think assuming non-empty is OK _given that `getContainerConfiguration()` is not null_.`

That said, I noticed `getContainerConfiguration()` is `@Nullable`. I think we should figure out when this can be null and why.

Lastly, using streams works, but for just getting the first element, I'd go with
```java
    platforms.iterator().next();
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 2]
```diff
@@ -260,14 +265,22 @@ private Image pullBaseImage(
...
+      Platform platform = platforms.stream().findFirst().get();
+      architecture = platform.getArchitecture();
+      os = platform.getOs();
+    }
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:281]: I find no reason to declare these as method fields.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 1 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:277]: > That said, I noticed `getContainerConfiguration()` is `@Nullable`. I think we should figure out when this can be null and why.

See #2618 and #2619.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 2 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:281]: Thats the only way we can access the variable inside the platform object.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 1 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:277]: Once #2619 is merged ,then I will need to adjust my code a little bit

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 3]
```diff
@@ -260,14 +265,22 @@ private Image pullBaseImage(
...
+
+    List<String> digests = manifestListTemplate.getDigestsForPlatform(architecture, os);
```
[the remark this answers is not in the document]
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:283]: Sounds Good I just added it

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 2 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:281]: What I meant is basically this:
```java
String architecture = "amd64"; // no need to have them as method fields, just local variables
String os = "linux";
if (...) {
  ...
  architecture = platform.getArchitecture();
  os = platforms.getOs();
}
List<String> digests = manifestListTemplate.getDigestsForPlatform(architecture, os);
```

Even with this, it's still not ideal because this `PullBaseImageStep` is yet another entity having an opinion of what a default should be. But this issue will be gone once #2619 is merged.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 4]
```diff
@@ -260,22 +263,40 @@ private Image pullBaseImage(
...
 
-    List<String> digests = manifestListTemplate.getDigestsForPlatform("amd64", "linux");
+    Set<Platform> platforms = buildContext.getContainerConfiguration().getPlatforms();
+    Platform platform = platforms.iterator().next();
+    String architecture = platform.getArchitecture();
+    String os = platform.getOs();
+
+    buildContext
+        .getEventHandlers()
+        .dispatch(
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:281]:
```suggestion
    EventHandlers eventHandlers = buildContext.getEventHandlers();
    eventHandlers.dispatch(
```

Then on line 300, you can reuse `eventHandlers`.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 5]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
 
     Mockito.verify(buildContext, Mockito.never()).newBaseImageRegistryClientFactory();
   }
+
```
Comment (louismurerwa) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:141]: i'm skeptical about my implementation for this testing method .

Basically its 

1.  Pass in a mock manifest list to the `testObtainPlatformSpecificImageManifest` method.
2.  Method searches for the digest matching the input platform
3.  Method calls registry client with with the obtained digest to get the mock manifest

Looking for feedback

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 6]
```diff
@@ -260,22 +263,40 @@ private Image pullBaseImage(
...
-              + " If your intention was to use a non-amd64/linux base image,"
+              + " is a manifest list, but the list does not contain an image manifest for the platform arch = "
+              + architecture
+              + ", os = "
+              + os
+              + ". If your intention was to specify a platform for your image,"
               + " see https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#how-do-i-specify-a-platform-in-the-manifest-list-or-oci-index-of-a-base-image"
-              + " to learn how to specify a manifest instead of a manifest list, until Jib fixes"
-              + " https://github.com/GoogleContainerTools/jib/issues/1567 to allow specifying architecture and OS.";
+              + " to learn more about specifyng a platform using the platform tag";
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:298]: "tag" is a confusing term, so let's just skip it.
```suggestion
              + " to learn more about specifying a platform";
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 7]
```diff
@@ -260,22 +263,40 @@ private Image pullBaseImage(
...
+
+    List<String> digests = manifestListTemplate.getDigestsForPlatform(architecture, os);
     if (digests.size() == 0) {
       String errorMessage =
           buildContext.getBaseImageConfiguration().getImage()
-              + " is a manifest list, but the list does not contain an image manifest for amd64/linux."
-              + " If your intention was to use a non-amd64/linux base image,"
+              + " is a manifest list, but the list does not contain an image manifest for the platform arch = "
+              + architecture
+              + ", os = "
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:294]:
```suggestion
              + " is a manifest list, but the list does not contain an image manifest for the platform architecture="
              + architecture
              + ", os="
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 8]
```diff
@@ -260,22 +263,40 @@ private Image pullBaseImage(
...
+    String architecture = platform.getArchitecture();
+    String os = platform.getOs();
+
+    buildContext
+        .getEventHandlers()
+        .dispatch(
+            LogEvent.lifecycle(
+                "Searching the manifest list for the platfrom, architecture =  "
+                    + architecture
+                    + ", os = "
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:285]:
```suggestion
            LogEvent.lifecycle(
                "Searching the manifest list for the platform, architecture="
                    + architecture
                    + ", os="
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 9]
```diff
@@ -260,22 +263,40 @@ private Image pullBaseImage(
...
+   * the first manifest it finds.
    */
-  private ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
+  @VisibleForTesting
+  ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
       RegistryClient registryClient, V22ManifestListTemplate manifestListTemplate)
       throws RegistryException, IOException {
 
-    List<String> digests = manifestListTemplate.getDigestsForPlatform("amd64", "linux");
+    Set<Platform> platforms = buildContext.getContainerConfiguration().getPlatforms();
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:274]: I think we need to throw an error if there is more than one platform. Something along the line of

```
Preconditions.checkArgument(platforms.size() == 1, "multiple base image platforms is not yet support; please specify only one platform");
```

You can do this in a separate PR if you want. Regardless, you need to add a test.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 10]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
...
+            + "    }\n"
+            + "}";
+    Mockito.when(buildContext.getContainerConfiguration()).thenReturn(containerConfiguration);
+    Mockito.when(containerConfiguration.getPlatforms())
+        .thenReturn(ImmutableSet.of(new Platform("testArchitecture", "testOS")));
+    V22ManifestListTemplate testManifestList =
+        JsonTemplateMapper.readJson(testManifestListTemplate, V22ManifestListTemplate.class);
+
+    V22ManifestTemplate testManifest =
+        new ObjectMapper().readValue(testManifestTemplate, V22ManifestTemplate.class);
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:189]: Be consistent and use either `JsonTemplateMapper` or `new ObjectMapper()` across the file.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 11]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
 
     Mockito.verify(buildContext, Mockito.never()).newBaseImageRegistryClientFactory();
   }
+
+  @Test
+  public void testObtainPlatformSpecificImageManifest()
+      throws IOException, InterruptedException, GeneralSecurityException, URISyntaxException,
+          RegistryException {
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:145]:
```suggestion
  public void testObtainPlatformSpecificImageManifest() throws IOException, RegistryException {
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 12]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
...
+            + "            \"os\": \"testOS\"\n"
+            + "         }\n"
+            + "      },\n"
+            + "      {\n"
+            + "         \"mediaType\": \"application/vnd.docker.distribution.manifest.v2+json\",\n"
+            + "         \"size\": 425,\n"
+            + "         \"digest\": \"sha256:5bb8e50aa2edd408bdf3ddf61efb7338ff34a07b762992c9432f1c02fc0e5e62\",\n"
+            + "         \"platform\": {\n"
+            + "            \"architecture\": \"Architecture\",\n"
+            + "            \"os\": \"OS\"\n"
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:166]:
```suggestion
            + "            \"architecture\": \"targetArchitecture\",\n"
            + "            \"os\": \"targetOS\"\n"
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 13]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
...
+            + "    \"mediaType\": \"application/vnd.docker.distribution.manifest.v2+json\",\n"
+            + "    \"config\": {\n"
+            + "        \"mediaType\": \"application/vnd.docker.container.image.v1+json\",\n"
+            + "        \"size\": 424,\n"
+            + "        \"digest\": \"sha256:5bb8e50aa2edd408bdf3ddf61efb7338ff34a07b762992c9432f1c02fc0e5e62\"\n"
+            + "    }\n"
+            + "}";
+    Mockito.when(buildContext.getContainerConfiguration()).thenReturn(containerConfiguration);
+    Mockito.when(containerConfiguration.getPlatforms())
+        .thenReturn(ImmutableSet.of(new Platform("testArchitecture", "testOS")));
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:184]: I think it's more meaningful to make the code return the second manifest from the list.
```suggestion
        .thenReturn(ImmutableSet.of(new Platform("targetArchitecture", "targetOS")));
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 14]
```diff
@@ -127,4 +138,80 @@ public void testCall_offlineMode_cached()
...
+    V22ManifestListTemplate testManifestList =
+        JsonTemplateMapper.readJson(testManifestListTemplate, V22ManifestListTemplate.class);
+
+    V22ManifestTemplate testManifest =
+        new ObjectMapper().readValue(testManifestTemplate, V22ManifestTemplate.class);
+
+    ManifestAndDigest<?> manifestAndDigest =
+        new ManifestAndDigest(testManifest, Digests.computeJsonDigest(testManifest));
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:192]:
```suggestion
        new ManifestAndDigest<>(testManifest, Digests.computeJsonDigest(testManifest));
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java (hunk 5 above)
Comment (chanseokoh) [reply jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:141]: Yeah, think about what you are testing. Most of the assertions that verifies the manifest content is unnecessary, because what you are testing is whether the method can find the digest to pass to `registryClient.pullManifest()` and whether it returns the same manifest returned by `registryClient`. `registryClient.pullManifest` doesn't even need to return an actual manifest; a mock will work. I can talk about this later.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 15]
```diff
@@ -260,23 +263,40 @@ private Image pullBaseImage(
...
+    Platform platform = platforms.iterator().next();
+    String architecture = platform.getArchitecture();
+    String os = platform.getOs();
+
+    EventHandlers eventHandlers = buildContext.getEventHandlers();
+    eventHandlers.dispatch(
+        LogEvent.info(
+            "Searching the manifest list for the platform, architecture="
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:282]: So, perhaps

LogEvent.lifecycle("The base image reference is manifest list, searching for architecture=...")

(Probably "for the platform" is redundant too. Generally speaking, program/tool messages should be concise.)

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java [hunk 16]
```diff
@@ -127,4 +132,49 @@ public void testCall_offlineMode_cached()
...
+    ManifestAndDigest<?> manifest = Mockito.mock(ManifestAndDigest.class);
+    Mockito.<ManifestAndDigest<?>>when(
+            registryClient.pullManifest(
+                "sha256:222222222222222222222222222222222222222222222222222222222222222222"))
+        .thenReturn(manifest);
+
+    ManifestAndDigest<?> returnManifest =
+        pullBaseImageStep.obtainPlatformSpecificImageManifest(registryClient, manifestList);
+
+    Assert.assertEquals(manifest, returnManifest);
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:178]: nit: `assertEquals` may work, but we can use `assertSame` to check if two object references are same (i.e., if `manifest == returnManifest` evaluates to true, meaning the references are pointing to the same single instance in the memory). OTOH, `assertEquals` may return when two (possible separate) object instances are deemed "equal" (whose definition is given by the class implementation of `.equals()`) even if `manifest != returnManifest`.

Comment (louismurerwa) [reply jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStepTest.java:178]: Good point !! Object comparison requires ` .equals() `

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 9 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: @chanseokoh I am going to add this ` check `  in this PR

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:272]: Let's do it in a separate PR. We may choose a different place for the check.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:272]: Sorry, let's not* do it.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:272]: Oh, wait. What I said was right. Let's not do it here but in a separate PR.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java [hunk 17]
```diff
@@ -260,23 +261,43 @@ private Image pullBaseImage(
...
-  private ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
+  @VisibleForTesting
+  ManifestAndDigest<?> obtainPlatformSpecificImageManifest(
       RegistryClient registryClient, V22ManifestListTemplate manifestListTemplate)
-      throws RegistryException, IOException {
+      throws IOException, RegistryException {
 
-    List<String> digests = manifestListTemplate.getDigestsForPlatform("amd64", "linux");
+    Set<Platform> platforms = buildContext.getContainerConfiguration().getPlatforms();
+    Preconditions.checkArgument(
```
Comment (louismurerwa) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:273]: I added a new `check` here after you approved the pr

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java (hunk 9 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:272]: Ok let me revert my commit

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: Sounds good. Avoid significantly changing the original intention of a PR (e.g., adding a new feature that can be done in a separate PR) (unless necessary) when people started reviewing.

For the check, I'm thinking maybe we can check and raise an error at an earlier point. We should also add a test too anyway.

Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: Yes the feature requires a test or two to confirm its functionality .At which point ,do you think we could do the check?

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: `JibContainerBuilder.containerize()` came to my mind, but I haven't given much thought actually. I'll need to look into the code, but in the meantime, you can continue work on other stuff.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: I think we can validate the input in `JibContainerBuilder.containerize()` and that `UnsupportedOperationException` could work. But if you think there's a better place, we can discuss.

```java
    try (BuildContext buildContext = toBuildContext(containerizer);
        TimerEventDispatcher ignored =
            new TimerEventDispatcher(
                buildContext.getEventHandlers(), containerizer.getDescription())) {
      if (...) {
        throw UnsupportedOperationException("... not yet supported ...");
      }
      logSources(buildContext.getEventHandlers());
```

Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java:271]: Single Platform Specification Support #2641
